reference/constraints/When.rst

```diff
@@ -0,0 +1,202 @@
+When
+===========
+
+This constraint allows you to apply constraints validation only if the provided expression returns true.
+See `Basic Usage`_ for an example.
+
+==========  ===================================================================
+Applies to  :ref:`class <validation-class-target>`
+            or :ref:`property/method <validation-property-target>`
+Options     - :ref:`expression <reference-constraint-expression-option>`
+            - :ref:`constraints <reference-constraint-constraints-option>`
+            - `groups`_
+            - `payload`_
+            - `values`_
+Class       :class:`Symfony\\Component\\Validator\\Constraints\\When`
+Validator   :class:`Symfony\\Component\\Validator\\Constraints\\WhenValidator`
+==========  ===================================================================
+
+Basic Usage
+-----------
+
+Imagine you have a class ``Discount`` with ``type`` and ``value``
+properties::
+
+    // src/Model/Discount.php
+    namespace App\Model;
+
+    use Symfony\Component\Validator\Constraints as Assert;
+
+    class Discount
+    {
+        private $type;
+
+        private $value;
+
+        // ...
+
+        public function getType()
+        {
+            return $this->type;
+        }
+
+        public function getValue()
+        {
+            return $this->value;
+        }
+
+        // ...
+    }
+
+To validate the object, you have some requirements:
+
+A) If ``type`` is ``percent``, then ``value`` must be less than 100;
+B) If ``type`` is ``absolute``, then ``value`` can be anything;
+C) No matter the value of ``type``, the ``value`` must be greater than 0.
+
+One way to accomplish this is with the When constraint:
+
+.. configuration-block::
+
+    .. code-block:: php-annotations
+
+        // src/Model/Discount.php
+        namespace App\Model;
+
+        use Symfony\Component\Validator\Constraints as Assert;
+
+        class Discount
+        {
+            /**
+             * @Assert\GreaterThan(0)
+             * @Assert\When(
+             *     expression="this.type == 'percent'",
+             *     constraints={@LessThan(100, message="The value should be between 0 and 100!")}
```

Comment:
what about juste writing `The value should be less than 100`
as it is the `@Assert\GreaterThan(0)` that validates the 0

Reply:
It's just an example and in my opinion, this message is more user friendly.